docs/Testing-GitHub-Actions.md

```diff
@@ -0,0 +1,142 @@
+
+## Test Types
+
+Tests for actions can be categorized into two broad categories.
+
+**Unit tests** are used to validate the business logic like - string manipulations, for loop, processing platform call responses. Unit tests are not expected to trigger the platform calls like making azure CLI / REST calls. Alternatively, developers can mock such platform calls. These tests can be run as part of build steps without requiring any environment setup. Developers can also use them during the development phase.
+
+Unit tests should focus on good code coverage. Every checkin should add tests to cover newly added code. Target is to keep code coverage at 80% and it will be made part of the PR approval checklist.
+
+**Functional tests**** (** also known as end to end tests ) validates end to end functional scenarios. It is important to have a handful of such tests to ensure functionality end to end. A functional test should be designed to cover the real time scenario. Our recommended way to create functional test is to create workflow and execute action with different input, scenario combinations.
+
+## Running Tests:
+
+Once suites of unit tests and function tests are ready, they will be run in different life-cycles of the development process.
+
+#### During Action Authoring:
+
+- Running Unit test using VS Code / VS IDE ( Details to be added )
+- Running functional test from dev box ( Details to be added )
+
+#### Pull Request
+
+During pull request, we want to ensure action logic is tested using unit tests on all platform matrices and there is required code coverage. We have designed the YAML to ensure these coverages. Templates for actions will be set up to run the unit tests for the master and release branches.
+
+- Yaml file for PR is configured to do the following:
+  - Run on pull requests in master and release branches.
+  - This workflow should run on all three OS platforms - windows, ubuntu, macos. One can make use of [matrix strategy](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/managing-complex-workflows#using-a-build-matrix) for the same.
+  - Workflow should have a build step for syntax validations and followed by a step to run the test.
+  - It is also recommended to add the test coverage tooling with your infra. The test coverage should provide details for newly added code like - code coverage percentage, file wise code coverage details.
+
+- Pull request reviewer must check for following:
+  - Code reviewers should be able to see code coverage data before approving the PR. PR approval must be done only when code coverage is more than 80%. To start with, code coverage should be measured at file level, which is each file has 80% coverage.
+
+Developer is expected to maintain the test suite copy with each branch to maintain the product code to test code compatibility.
+
+## Releasing Action:
```

Comment:
Should this be - "Validation before Release" ?

Reply:
@ds-ms suggestion